include/bluetooth/audio/hap.h

```diff
@@ -0,0 +1,275 @@
+	 *
+	 * This callback notifies the application of a new profile connection.
+	 * In case the err parameter is non-zero it means that the profile
+	 * connection failed.
+	 *
+	 * @param hap Profile instance object if any.
+	 * @param ha_type Hearing Aid Type.
+	 * @param error Zero for success, non-zero otherwise.
+	 */
+	void (*connected)(struct bt_hap *hap, uint8_t ha_type, int error);
+
+	/**
+	 * @brief A profile has been disconnected.
+	 *
+	 * This callback notifies the application that a profile
+	 * has been disconnected.
+	 *
+	 * @param hap Profile instance object.
+	 * @param reason Reason for the disconnection.
+	 */
+	void (*disconnected)(struct bt_hap *hap, int reason);
+
+	/**
+	 * @brief A profile instance has been deleted.
+	 *
+	 * This callback notifies the application that a profile
+	 * instance has been removed. This may be called when the bond has
+	 * been removed or @ref bt_hap_delete has been called.
+	 *
+	 * @param hap Profile instance object.
+	 */
+	void (*deleted)(struct bt_hap *hap);
+
+	/**
+	 * @brief An active preset has been changed.
+	 *
+	 * This callback notifies the application when the active preset or
+	 * it's name has been changed.
+	 * If @kconfig{BT_HAP_HARC_PRESET_CACHE_SIZE} is 0 the preset name will
+	 * not be set, but if @kconfig{BT_HAP_HARC_AUTO_READ_ACTIVE_PRESET}
+	 * option is enabled, the callback will be called again on preset read
+	 * complete.
+	 *
+	 * @param hap Profile instance object.
+	 * @param preset Preset record object.
+	 */
+	void (*active_preset_changed)(struct bt_hap *hap,
+				      const struct bt_has_preset *preset);
+
+	/**
+	 * @brief A cached preset list has been updated.
+	 *
+	 * This callback notifies the application when the list of preset
+	 * records changed i.e. record has been added/removed, changed name or
+	 * become available.
+	 * The user can call @ref bt_hap_preset_foreach to obtain the current
+	 * presets.
+	 *
+	 * @param hap Profile instance object.
+	 */
+	void (*preset_cache_updated)(struct bt_hap *hap);
+};
+
+/**
+ * @def BT_HAP_CB_DEFINE
+ *
+ * @brief Register a callback structure for Hearing Access Profile events.
+ *
+ * @param _name Name of callback structure.
+ */
+#define BT_HAP_CB_DEFINE(_name)							\
+	static const STRUCT_SECTION_ITERABLE(bt_hap_cb,				\
+					     _CONCAT(bt_hap_cb_, _name))
+
+/**
+ * @brief Create profile instance to remote Hearing Access Service
+ *
+ * This will start a GATT discovery to setup handles and subscriptions.
+ * This shall be called once before any other actions can be executed for the
+ * peer device, and the @ref bt_hap_cb connected callback will notify
+ * when it is possible to start remote operations.
+ * The profile instance is removed when the bond is removed or when @ref
+ * bt_hap_delete is called.
+ * 
+ * @param conn The connection to initialize the profile for.
+ *
+ * @return 0 in case of success or negative value in case of error.
+ */
+int bt_hap_new(struct bt_conn *conn);
```

Comment:
In the case of being connect to 2 or more devices (2 being quite likely) this function does not give the user a proper relationship between the `bt_conn` and the `bt_hap` pointers. If a device does `bt_hap_new` on both `bt_conn` at the same time, it will not be able to determine which `connected` callback is for which `bt_conn`. 

Suggest to either return the `bt_hap` as an output pointer here (has been used for several other LE audio implementations), or to add the `bt_conn` to the `connected` callback.

Reply:
The user can use `bt_hap_conn_get` to get the context, but I'm OK with having `bt_conn` in `connected` callback parameters.

I have other doubts, regarding using `bt_conn` in this API. Instead `bt_conn` I would like to use some abstraction to group of devices (coordinated set)

Reply:
> The user can use bt_hap_conn_get to get the context, but I'm OK with having bt_conn in connected callback parameters.

Good point, that should actually probably be enough. 

> I have other doubts, regarding using bt_conn in this API. Instead bt_conn I would like to use some abstraction to group of devices (coordinated set)

Good point, it should probably use the struct from the updated set coordinator API. 

I do wonder (perhaps that is defined by one of the profiles): If you are expecting to connect to a hearing aid, would you discover CSIS or HAS first? I assume that you will only connect to those advertising the right UUIDs, but is there a defined order of discovery procedures? 

And in the context of HAP, where the HA might do local synchronization, should you then do the HAP discovery procedure, and then see if it is locally synchronized, before discovering the CSIS instance? 
If you only intend on setting presets via BT, then you possible never need to do CSIP discovery. 

(in any case the CSIP member struct works without discovering CSIS).